cromwell_tools/_cromwell_api.py

```diff
@@ -0,0 +1,425 @@
+            ValueError: If a list of values are assigned to a query key that belongs to _cromwell_exclusive_query_keys.
+        Returns:
+            query_params (List[dict]): A composed list of query objects.
+        """
+        if not isinstance(query_dict, dict):
+            raise TypeError('A valid dictionary with query keys is required!')
+
+        query_params = []
+        for k, v in query_dict.items():
+            if k in cls._cromwell_query_keys:
+                if k is 'label' and isinstance(v, dict):
+                    query_params.extend([
+                        {'label': label_key + ':' + label_value} for label_key, label_value in v.items()
+                    ])
+                elif isinstance(v, list):
+                    if k in cls._cromwell_exclusive_query_keys:
+                        raise ValueError('{} cannot be specified multiple times!'.format(k))
+                    query_params.extend([
+                        {k: json.dumps(val)} if not isinstance(val, str) else {k: val} for val in set(v)
+                    ])
+                else:
+                    query_params.append(
+                            {k: json.dumps(v)} if not isinstance(v, str) else {k: v}
+                    )
+            else:
+                logger.info('{} is not an allowed query key in Cromwell, will be ignored in this query.'.format(k))
+        return query_params
+
+    @staticmethod
+    def _parse_status(response):
+        """Helper function to parse a status response.
+
+        Args:
+            response (requests.Response): A status response object from Cromwell.
+
+        Raises:
+            WorkflowUnknownException: This will be raised when Cromwell returns a status code != 200.
+
+        Returns:
+            str: String representing status response.
+        """
+        if response.status_code != 200:
+            raise WorkflowUnknownException(
+                    'Status could not be determined, endpoint returned {0}'.format(
+                            response.status_code))
+        else:
+            return response.json()['status']
+
+    @staticmethod
+    def _check_and_raise_status(response):
+        """Helper function to check the status of a response and raise a friendly message if there are errors.
+
+        This functions is using the `response.ok` which wraps the `raise_for_status()`, by doing this, we can
+        produce the actual error messages from the Cromwell, instead of shadowing them with `raise_for_status()`.
+
+        Args:
+            response (requests.Response): A status response object from Cromwell.
+
+        Raises:
+            requests.exceptions.HTTPError: This will be raised when Cromwell returns a response that satisfies
+                400 <= response.status_code < 600.
+        """
+        if not response.ok:
+            raise requests.exceptions.HTTPError('Error Code {0}: {1}'.format(response.status_code, response.text))
```

Comment:
Can you remind me why `_check_and_raise_status` is now configurable? When would users want to use one option over another?

Reply:
So in general, there are 2 use cases:
- Users(such as Lira) wants to have their own retry policy based on the response code, and they might also want to deal with the response body directly. In these cases, this library should give them the raw/transparent response from Cromwell. (_check_and_raise_status=False by default)
- There are other situations that users(such as who invoke this function from CLI) prefer to know more about the error message than dealing with the errors, and they don't care about who raises the error(Cromwell or cromwell-tools), as long as the error messages are friendly and reasonable. In these cases, we will need cromwell-tools and catch and raise the errors in a proper way. (_check_and_raise_status=True)